### DISCUSSION TITLE
[ImportError] ModuleNotFoundError: No module named 'socketio.exceptions'

### DISCUSSION THREAD

**Discussion by crhodes2** [miguelgrinberg/Flask-SocketIO#1679] [Q&A]
All I am trying to do is have the flask-socketio module installed and running onto my flask server application.
I have done the pip install, and tried to plug in the import and I keep getting this error.` **from flask_socketio import SocketIO** `
and as a result, I keep getting this error.
```
Traceback (most recent call last):
  File "app.py", line 10, in <module>
    from flask_socketio import SocketIO
  File "/usr/local/lib/python3.7/dist-packages/flask_socketio/__init__.py", line 22, in <module>
    from socketio.exceptions import ConnectionRefusedError  # noqa: F401
ModuleNotFoundError: No module named 'socketio.exceptions'

```
Is this how the module supposed to work? How can I move away past this? I have recreated possible solutions online using the following links, but the issue refuses to go away.
https://github.com/miguelgrinberg/Flask-SocketIO/issues/1105

Whoever build this module, I'm sorry to say but I absolutely abhor your module. It has done nothing but drive me crazy all week trying to run this to my Flask server, I am debating about moving away from it.

**Comment by miguelgrinberg** [comment]
> I'm sorry to say but I absolutely abhor your module

This is not a good way to ask for help. The person who built this module is the person that will try to help you, so yeah, I get that you are frustrated, but without offending you, you have to consider that thousands of users use this package, so this is likely your own user error.

Also, if I may complain myself as well, you have written an "issue" on this repository, while in fact you don't know if there is an issue or if the issue is yours. Issues are for bugs that I need to work on. If you have a question, need help, or don't know where the issue is, the proper thing to do is to open a discussion. I'm moving this to a discussion now.

No, on to your issue: My guess is that you have installed a set of dependencies that are incompatible with each other. So start by upgrading flask-socketio to the latest version with this command:

    pip install --upgrade flask-socketio

This should bring everything to the latest versions. If you continue to have errors after this, let me know.

**Reply by crhodes2** [reply]
Well this is not some user error on my part. If there was then it should've been in situations if I have installed some other socketio dependencies into my requirements. But in that case I did not. I tried recreating my flask application project from scratch, included blueprints, __ init __ files, and all, and installed the flask-socketio application by itself, without adding any other modules I have added before on my requirements.txt; and despite that, I kept getting the error.

Well, I solved this issue. Unfortunately, your solution was the first thing I have tried when I started having that error and I was not getting any successes. What I did instead was upgrading setuptools; coming from my seventh stack overflow search: https://stackoverflow.com/questions/44761958/using-pip3-module-importlib-bootstrap-has-no-attribute-sourcefileloader

I'm sorry if I have ever offended you over this module but after spending a week trying and failing to import this module into my project, throwing the towel was my only option. And for someone who doesn't know when to quit, this was a painful, excruciating moment for me and I do not want to relive that again.

**Reply by miguelgrinberg** [reply]
@crhodes2 glad to hear you figured out the problem.

How did you make the leap from a "module not found" error to this Stack Overflow question, which has absolutely no relation to your error?

> Well this is not some user error on my part

My point still stands. This package is actively worked on, maintained, and with a large user base. On the other side, your system was unable to properly install packages from PyPI. A clean installation of Python 3.7 would not have this problem (I know, I still use Python 3.7 on my Chromebook all the time, and a large part of the code in this package was written on it). Something must have happened to your 3.7 installation that put it into this bad state. Nothing in this package caused your problem, so I hope we agree that you were wrong in "abhoring" this package, as it clearly had nothing to do with your problem. And I hope you'll reconsider how to approach a maintainer of an open source project when seeking help.

**Reply by crhodes2** [reply]
"How did you make the leap from a "module not found" error to this Stack Overflow question, which has absolutely no relation to your error?" well, the error was looking for socketio.exception, which was missing, so I've tried installing that and it gave me a different error, so using that stack overflow error forum, I solved the situation. This issue was basically taking me to a labyrinths of errors after the next.

I'm sorry to say but I am not going to agree that it was an user error and was wrong on this matter. Let's just leave it at that.

**Reply by miguelgrinberg** [reply]
> This issue was basically taking me to a labyrinths of errors after the next.

Right. And somehow you thought it was okay to just include the first of those labyrinth of errors here and blame this package for your troubles, without providing the whole story.  And even now that you have solved the problem, you do not provide all the steps for other users behind you that may have the same problem and can benefit from your investigation and solution. Is it also my fault that you are terrible at reporting issues, or can you at least concede that one?

At this point I really wish you good luck and I hope it is a smooth sailing until you complete your project, without the need for additional help. Because I hope you will understand if I prefer to spend my time helping users who appreciate the work that I do.
